libraries/AP_Proximity/AP_Proximity_MAV.cpp

```diff
@@ -44,132 +45,208 @@ bool AP_Proximity_MAV::get_upward_distance(float &distance) const
+        if (database_ready) {
+            database_push(mid_angle, packet_distance_m, _last_update_ms, current_pos, body_to_ned);
+        }
+    }
 
-            const float packet_distance_m = distance_cm * 0.01f;
-            const float mid_angle = wrap_360((float)j * increment + yaw_correction);
-
-            // iterate over proximity sectors
-            for (uint8_t i = 0; i < PROXIMITY_NUM_SECTORS; i++) {
-                // update distance array sector with shortest distance from message
-                const float angle_diff = wrap_180(_sector_middle_deg[i] - mid_angle);
-                if (fabsf(angle_diff) > PROXIMITY_SECTOR_WIDTH_DEG*0.5f) {
-                    // not even in this sector
-                    continue;
-                }
-                if (is_equal(angle_diff, -PROXIMITY_SECTOR_WIDTH_DEG*0.5f)) {
-                    // on the upper boundary is *out* to avoid
-                    // ambiguity.  The distance is considered to be in
-                    // the next sector.  We should never be within an
-                    // epsilon of the boundary, so is_equal should be
-                    // safe.
-                    continue;
-                }
-                if (packet_distance_m >= _distance[i]) {
-                    // this is no closer than a previous distance
-                    // found from the pakcet
-                    continue;
-                }
-
-                // this is the shortest distance we've found in the packet so far
-                _distance[i] = packet_distance_m;
-                _angle[i] = mid_angle;
-                sector_updated[i] = true;
-            }
+    // process the last face
+    if (face_distance_valid) {
+        boundary.set_face_attributes(face, face_yaw_deg, face_distance);
+    } else {
+        boundary.reset_face(face);
+    }
+    return;
+}
 
-            // update Object Avoidance database with Earth-frame point
-            if (database_ready) {
-                database_push(mid_angle, packet_distance_m, _last_update_ms, current_pos, body_to_ned);
-            }
-        }
+// handle mavlink OBSTACLE_DISTANCE_3D messages
+void AP_Proximity_MAV::handle_obstacle_distance_3d_msg(const mavlink_message_t &msg)
+{
+    mavlink_obstacle_distance_3d_t packet;
+    mavlink_msg_obstacle_distance_3d_decode(&msg, &packet);
 
-        // update proximity sectors validity and boundary point
-        for (uint8_t i = 0; i < PROXIMITY_NUM_SECTORS; i++) {
-            _distance_valid[i] = (_distance[i] >= _distance_min) && (_distance[i] <= _distance_max);
-            if (sector_updated[i]) {
-                update_boundary_for_sector(i, false);
-            }
-        }
+    const uint32_t previous_sys_time = _last_update_ms;
+    _last_update_ms = AP_HAL::millis();
+   
+    // time_diff will check if the new message arrived significantly later than the last message
+    const uint32_t time_diff = _last_update_ms - previous_sys_time;
```

Comment:
This will work OK if a sender keeps on sending a distance that falls within the same face at a very high rate?  Imagine that the vehicle is not moving and there's an obstacle directly in front of the vehicle and the sender has a really fast update rate.  I think the boundary should be updated every PROXIMITY_3D_MSG_TIMEOUT_MS even though the sender is sending in data faster than this.